crates/matrix-sdk-ui/src/timeline/inner/state.rs

```diff
@@ -444,6 +258,188 @@ impl DerefMut for TimelineInnerState {
+            is_own_event,
+            encryption_info: event.encryption_info,
+            read_receipts: if settings.track_read_receipts {
+                self.load_read_receipts_for_event(&event_id, room_data_provider).await
+            } else {
+                Default::default()
+            },
+            is_highlighted: event.push_actions.iter().any(Action::is_highlight),
+            flow: Flow::Remote { event_id, raw_event: raw, txn_id, position, should_add },
+        };
+
+        TimelineEventHandler::new(self, ctx, settings.track_read_receipts).handle_event(event_kind)
+    }
+
+    pub(super) fn clear(&mut self) {
+        // By first checking if there are any local echoes first, we do a bit
+        // more work in case some are found, but it should be worth it because
+        // there will often not be any, and only emitting a single
+        // `VectorDiff::Clear` should be much more efficient to process for
+        // subscribers.
+        if self.items.iter().any(|item| item.is_local_echo()) {
+            // Remove all remote events and the read marker
+            self.items.for_each(|entry| {
+                if entry.is_remote_event() || entry.is_read_marker() {
+                    ObservableVectorTransactionEntry::remove(entry);
+                }
+            });
+
+            // Remove stray day dividers
+            let mut idx = 0;
+            while idx < self.items.len() {
+                if self.items[idx].is_day_divider()
+                    && self.items.get(idx + 1).map_or(true, |item| item.is_day_divider())
+                {
+                    self.items.remove(idx);
+                    // don't increment idx because all elements have shifted
+                } else {
+                    idx += 1;
+                }
+            }
+        } else {
+            self.items.clear();
+        }
+
+        self.reactions.clear();
+        self.fully_read_event = None;
+        self.event_should_update_fully_read_marker = false;
+    }
+
+    #[instrument(skip_all)]
+    pub(super) fn set_fully_read_event(&mut self, fully_read_event_id: OwnedEventId) {
+        // A similar event has been handled already. We can ignore it.
+        if self.fully_read_event.as_ref().is_some_and(|id| *id == fully_read_event_id) {
+            return;
+        }
+
+        self.fully_read_event = Some(fully_read_event_id);
+        self.meta.update_read_marker(&mut self.items);
+    }
+}
+
+impl Deref for TimelineInnerStateTransaction<'_> {
+    type Target = TimelineInnerMetadata;
+
+    fn deref(&self) -> &Self::Target {
+        self.meta
+    }
+}
+
+impl DerefMut for TimelineInnerStateTransaction<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.meta
+    }
+}
+
+impl Drop for TimelineInnerStateTransaction<'_> {
+    fn drop(&mut self) {
+        // Safety: self.items is not touched again
+        let txn = unsafe { ManuallyDrop::take(&mut self.items) };
+        txn.commit();
```

Comment:
Is the `ManuallyDrop` needed because `commit()` takes ownership of self? If so, this might be useful to comment somewhere.